From f597c031b3d5236f598219fd960fb2cf4541f513 Mon Sep 17 00:00:00 2001
From: Jasper van Bourgognie <jasper@knoop.je>
Date: Mon, 15 Sep 2025 07:42:51 +0200
Subject: [PATCH] meson: build options for 3/4 finger dragging

Add an option to set the default 3FG drag behaviour.

Add an option no to disambiguate between swipe and drag in 3 finger
mode, but always go to drag
---
 meson.build                      | 13 +++++++++++++
 meson_options.txt                |  9 +++++++++
 src/evdev-mt-touchpad-gestures.c |  6 +++++-
 3 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 85493425..165e1877 100644
--- a/meson.build
+++ b/meson.build
@@ -411,6 +411,19 @@ if get_option('internal-event-debugging')
 	config_h.set('EVENT_DEBUGGING', 1)
 endif
 
+if get_option('3fg-drag-always-drag')
+	config_h.set('DRAG_3FG_ALWAYS_DRAG', 1)
+endif
+
+drag_3fg_default = get_option('3fg-drag-default')
+if drag_3fg_default == 'disabled'
+	config_h.set('DRAG_3FG_DEFAULT', 'LIBINPUT_CONFIG_3FG_DRAG_DISABLED')
+elif drag_3fg_default == '3fg'
+	config_h.set('DRAG_3FG_DEFAULT', 'LIBINPUT_CONFIG_3FG_DRAG_ENABLED_3FG')
+elif drag_3fg_default == '4fg'
+	config_h.set('DRAG_3FG_DEFAULT', 'LIBINPUT_CONFIG_3FG_DRAG_ENABLED_4FG')
+endif
+
 config_h.set_quoted('LIBINPUT_PLUGIN_LIBDIR', dir_lib / 'libinput' / 'plugins')
 config_h.set_quoted('LIBINPUT_PLUGIN_ETCDIR', dir_etc / 'libinput' / 'plugins')
 
diff --git a/meson_options.txt b/meson_options.txt
index 20d7e05e..4dc4f04e 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -50,3 +50,12 @@ option('lua-plugins',
 	type: 'feature',
 	value: 'auto',
 	description: 'Enable support for Lua plugins')
+option('3fg-drag-always-drag',
+	type: 'boolean',
+	value: false,
+	description: 'Skip the 3-finger drag-vs-swipe disambiguation timeout and always commit to a drag (default=false)')
+option('3fg-drag-default',
+	type: 'combo',
+	choices: ['disabled', '3fg', '4fg'],
+	value: '3fg',
+	description: 'Default value for the 3-finger drag config option (default=3fg)')
diff --git a/src/evdev-mt-touchpad-gestures.c b/src/evdev-mt-touchpad-gestures.c
index 03a081c0..18a8b3ec 100644
--- a/src/evdev-mt-touchpad-gestures.c
+++ b/src/evdev-mt-touchpad-gestures.c
@@ -1908,6 +1908,9 @@ tp_gesture_handle_state_3fg_drag_or_swipe(struct tp_dispatch *tp, usec_t time)
 static void
 tp_gesture_handle_state_3fg_drag_or_swipe_start(struct tp_dispatch *tp, usec_t time)
 {
+#ifdef DRAG_3FG_ALWAYS_DRAG
+	tp->gesture.state = GESTURE_STATE_3FG_DRAG_START;
+#else
 	const struct normalized_coords zero = { 0.0, 0.0 };
 	gesture_notify_swipe(&tp->device->base,
 			     time,
@@ -1917,6 +1920,7 @@ tp_gesture_handle_state_3fg_drag_or_swipe_start(struct tp_dispatch *tp, usec_t t
 			     &zero);
 	tp->gesture.state = GESTURE_STATE_3FG_DRAG_OR_SWIPE;
 	tp_gesture_set_3fg_drag_3fg_or_swipe_timer(tp, time);
+#endif
 }
 
 static void
@@ -2347,7 +2351,7 @@ tp_3fg_drag_get_enabled(struct libinput_device *device)
 static enum libinput_config_3fg_drag_state
 tp_3fg_drag_default(struct tp_dispatch *tp)
 {
-	return LIBINPUT_CONFIG_3FG_DRAG_DISABLED;
+	return DRAG_3FG_DEFAULT;
 }
 
 static enum libinput_config_3fg_drag_state
-- 
2.55.0

